src/main/java/org/xstefank/model/PersistentList.java

```diff
@@ -30,32 +34,136 @@ public PersistentList(String dirName, String fileName) {
+    public void add(int index, String element) {
+        super.add(index, element);
+        rewriteFileContent(index);
+    }
+
+    @Override
+    public String remove(int index) {
+        String element = super.remove(index);
+        rewriteFileContent(index);
+        return element;
+    }
+
+    @Override
+    public boolean remove(Object o) {
+        int index = indexOf(o);
+        boolean removed = super.remove(o);
+        if (removed) {
+            rewriteFileContent(index);
+        }
+        return removed;
+    }
+
+    @Override
+    public void clear() {
+        super.clear();
+        try (FileWriter fw = new FileWriter(file, false)) {
+            fw.write("");
+        } catch (IOException e) {
+            throw new IllegalStateException("Cannot delete list elements", e);
+        }
+    }
+
+    @Override
+    public boolean addAll(Collection<? extends String> c) {
+        boolean changed = super.addAll(c);
+        try (FileWriter fw = new FileWriter(file, true)) {
+            for (String element : c) {
+                fw.write(element);
+            }
+        } catch (IOException e) {
+            throw new IllegalStateException("Cannot write elements to list", e);
+        }
+        return changed;
+    }
+
+    @Override
+    public boolean addAll(int index, Collection<? extends String> c) {
+        boolean addedAll = super.addAll(index, c);
+        if (addedAll) {
+            rewriteFileContent(index);
+        }
+        return addedAll;
+    }
+
+    @Override
+    protected void removeRange(int fromIndex, int toIndex) {
+        super.removeRange(fromIndex, toIndex);
+        rewriteFileContent(fromIndex);
+    }
+
+    @Override
+    public boolean retainAll(Collection<?> c) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public boolean removeIf(Predicate<? super String> filter) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void replaceAll(UnaryOperator<String> operator) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void sort(Comparator<? super String> c) {
+        throw new UnsupportedOperationException();
     }
 
     private void loadFileContent(File file) {
         try (BufferedReader br = new BufferedReader(new FileReader(file))) {
             String line;
             while ((line = br.readLine()) != null) {
                 if (!line.startsWith("#")) {
-                    list.add(line);
+                    super.add(line);
                 }
             }
         } catch (IOException e) {
             throw new IllegalStateException("Cannot read file " + file.getPath(), e);
         }
     }
-}
+
+    private void rewriteFileContent(int fromIndex) {
+        try (RandomAccessFile randomAccessFile =
+                     new RandomAccessFile(file.getAbsolutePath(), "rw")) {
+            if (randomAccessFile.length() > 0) {
+                int fileEnd = 0;
```

Comment:
this should not be names as fileEnd. It is a temporal file end when you find fromIndex but the name is misleading.